src/main/java/com/stakater/nordmart/catalog/controller/CatalogController.java

```diff
@@ -61,4 +70,42 @@ public CatalogController(ProductRepository repository, MeterRegistry meterRegist
         productCount.set(productList.size());
         return productList;
     }
+
+    @ApiOperation(value = "Saves product.",
+            consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+    @PostMapping(consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+    public ProductDto saveProduct(@RequestBody ProductDto productDto) {
+        return catalogService.store(productDto);
+    }
+
+    @ApiOperation(value = "Updated product.",
+            consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+    @PutMapping(consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+    public ProductDto updateProduct(@RequestBody ProductDto productDto) {
+        return catalogService.update(productDto);
+    }
+
+    @ApiOperation(value = "Removes product by provided itemId. Returns 404 if product was not found.",
+            consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+    @DeleteMapping(path = "/{itemId}")
+    public ResponseEntity<String> deleteProduct(@PathVariable(name = "itemId") String itemId) {
+        try {
+            catalogService.delete(itemId);
+            return new ResponseEntity("Successfully deleted item with id " + itemId, HttpStatus.OK);
+        } catch (NotFoundException e) {
+            return new ResponseEntity("Item with id " + itemId + " was not found.", HttpStatus.NOT_FOUND);
+        }
+    }
+
+    @ApiOperation(value = "Gets product by provided itemId. Returns 404 status if such product  was not found.",
+            consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+    @GetMapping(path = "/{itemId}", produces = APPLICATION_JSON_VALUE)
+    public ResponseEntity getProduct(@PathVariable(name = "itemId") String itemId) {
+        Optional<ProductDto> productDto = catalogService.get(itemId);
+        if (productDto.isPresent()) {
+            return new ResponseEntity(productDto.get(), HttpStatus.OK);
+        } else {
+            return new ResponseEntity("Product with item id " + itemId + " not Found", HttpStatus.NOT_FOUND);
+        }
+    }
 }
```

Comment:
Ensure that these are added on front-end, handled in gateway and the functionality is working properly.

Reply:
Tested this functionality locally and also covered with integration test. Front-end is not part of my task.

Reply:
Hey, so you tested it using just the micro-services and didn't test it end-to-end ? Like you tested it locally without using front-end right ? Just want to know so that i am aware of what's left for us to do.

Reply:
Hi! I have tested search functionality on the system deployed to local Kubernetes cluster using UI, so basically end to end test. 
Also, I port forwarded catalog service API and used Postman to test CRUD operations of catalog service. And, of course, tested that all of them are correctly pushed to Kafka, then processed by search service, put to elasticsearch and can be searched after that.
So actual integration into the gateway, frontend, and e2e test using cypress left to be done from my perspective. This can be my next assignment if it makes sense to you.

Reply:
Thanks for the clarification, makes sense now.